compiler/AST/build.cpp

```diff
@@ -2185,6 +2185,180 @@ BlockStmt* buildLocalStmt(Expr* condExpr, Expr *stmt) {
+  auto seManager = new SymExpr(managerHandle);
+  auto enterThis = new CallExpr("enterThis", gMethodToken, seManager);
+
+  // BUILD: [var/ref/const ref] myResource = manager.enterThis();
+  if (resourceName != nullptr) {
+    const bool isResourceStorageKindInferred = (flags == nullptr);
+    auto resource = new VarSymbol(resourceName);
+
+    if (isResourceStorageKindInferred) {
+      resource->addFlag(FLAG_MANAGER_RESOURCE_INFER_STORAGE);
+    } else {
+      for (auto f: *flags) resource->addFlag(f);
+      delete flags;
+      flags = nullptr;
+    }
+
+    ret->insertAtTail(new DefExpr(resource, enterThis));
+
+  } else {
+
+    // Otherwise, just make the call to 'enterThis()'.
+    ret->insertAtTail(enterThis);
+  }
+
+  // BUILD: TEMP var errorCaught = false;
+  auto errorCaught = newTemp("errorCaught");
+  ret->insertAtTail(new DefExpr(errorCaught, new_BoolSymbol(false)));
+
+  // Call helper to construct try/catch block.
+  auto tryCatch = buildTryCatchForManagerBlock(managerHandle, errorCaught);
+  ret->insertAtTail(tryCatch);
+
+  // BUILD: if !errorCaught then manager.leaveThis(nil);
+  auto ifCond = new CallExpr(PRIM_UNARY_LNOT, new SymExpr(errorCaught));
+  auto ifBranch = new CallExpr("leaveThis",
+                               gMethodToken,
+                               new SymExpr(managerHandle),
+                               gNil);
+  auto ifStmt = new CondStmt(ifCond, ifBranch);
+  ret->insertAtTail(ifStmt);
+
+  return ret;
+}
+
+/*
+  Managers contains a list of blocks that have been prepared for each
+  manager expression. They are threaded into each other from left to
+  right, such that the Nth manager block contains the body of the N+1
+  block, etc. The final block is the user code. This naturally handles
+  resource allocation and cleanup.
+
+  TODO: In cleanup, recursively lift up the manager out of its try block
+        if we detect exception handling is not needed (e.g. we're not
+        in a throwing function, and not in a try).
+*/
+BlockStmt* buildManageStmt(BlockStmt* managers, BlockStmt* block) {
+  auto ret = new BlockStmt();
+
+  if (fWarnUnstable) {
+    USR_WARN(managers, "manage statements are not stable and may change");
+  }
+
+  // Used to thread context managers. Start by inserting into outer block.
+  BlockStmt* insertionPoint = ret;
+
+  for_alist(manager, managers->body) {
+    BlockStmt* managerBlock = toBlockStmt(manager);
+    INT_ASSERT(managerBlock);
+
+    bool hasInsertedManagerBlock = false;
+
+    for_alist(stmt, managerBlock->body) {
+      if (TryStmt* tryStmt = toTryStmt(stmt)) {
+
+        // Insert this manager block at the appropriate point.
+        managerBlock->remove();
+        insertionPoint->insertAtTail(managerBlock);
+        managerBlock->flattenAndRemove();
+        hasInsertedManagerBlock = true;
+
+        // The manager's try body is empty, insert there next.
+        insertionPoint = tryStmt->body();
+        break;
+      }
+    }
```

Comment:
We check for a `try` in the loop because the try body is the next insertion point. I will try and rephrase the comment.

It seems like your loop is pretty much mine except the insertion occurs before the next insertion point is found?

Reply:
Right right. And as I said, I'm ok with this implementation. But reading the code made me wonder whether there's more to it for this implementation then I thought.

Reply:
Swapped to your loop FWIW :)